textdb/textdb-dataflow/src/test/java/edu/uci/ics/textdb/dataflow/join/JoinDistanceTest.java

```diff
@@ -0,0 +1,1343 @@
+package edu.uci.ics.textdb.dataflow.join;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import edu.uci.ics.textdb.api.common.Attribute;
+import edu.uci.ics.textdb.api.common.FieldType;
+import edu.uci.ics.textdb.api.common.IField;
+import edu.uci.ics.textdb.api.common.ITuple;
+import edu.uci.ics.textdb.api.common.Schema;
+import edu.uci.ics.textdb.api.exception.TextDBException;
+import edu.uci.ics.textdb.common.constants.LuceneAnalyzerConstants;
+import edu.uci.ics.textdb.common.constants.SchemaConstants;
+import edu.uci.ics.textdb.common.constants.DataConstants.KeywordMatchingType;
+import edu.uci.ics.textdb.common.exception.DataFlowException;
+import edu.uci.ics.textdb.common.field.DataTuple;
+import edu.uci.ics.textdb.common.field.IntegerField;
+import edu.uci.ics.textdb.common.field.ListField;
+import edu.uci.ics.textdb.common.field.Span;
+import edu.uci.ics.textdb.common.field.StringField;
+import edu.uci.ics.textdb.common.field.TextField;
+import edu.uci.ics.textdb.common.utils.Utils;
+import edu.uci.ics.textdb.dataflow.common.FuzzyTokenPredicate;
+import edu.uci.ics.textdb.dataflow.common.JoinDistancePredicate;
+import edu.uci.ics.textdb.dataflow.fuzzytokenmatcher.FuzzyTokenMatcher;
+import edu.uci.ics.textdb.dataflow.keywordmatch.KeywordMatcherSourceOperator;
+import edu.uci.ics.textdb.dataflow.projection.ProjectionOperator;
+import edu.uci.ics.textdb.dataflow.projection.ProjectionPredicate;
+import edu.uci.ics.textdb.dataflow.source.IndexBasedSourceOperator;
+import edu.uci.ics.textdb.dataflow.utils.TestUtils;
+import edu.uci.ics.textdb.storage.DataStore;
+import edu.uci.ics.textdb.storage.relation.RelationManager;
+import junit.framework.Assert;
+
+/**
+ * 
+ * @author sripadks
+ * @author Zuozhi Wang
+ *
+ */
+public class JoinDistanceTest {
+    
+    public static final String BOOK_TABLE_OUTER = JoinDistanceHelper.BOOK_TABLE_OUTER;
+    public static final String BOOK_TABLE_INNER = JoinDistanceHelper.BOOK_TABLE_INNER;
+    
+    public static final KeywordMatchingType conjunction = KeywordMatchingType.CONJUNCTION_INDEXBASED;
+    public static final KeywordMatchingType phrase = KeywordMatchingType.PHRASE_INDEXBASED;
+
+    // writes the test tables before ALL tests
+    @BeforeClass
+    public static void setup() throws TextDBException {
+        JoinDistanceHelper.createTestTables();
+    }
+    
+    // deletes the test tables after ALL tests
+    @AfterClass
+    public static void cleanUp() throws TextDBException {
+        JoinDistanceHelper.deleteTestTables();
+    }
+    
+    // clear the test tables after EACH test
+    @After
+    public void clearTables() throws TextDBException {
```

Comment:
The `@After` annotation is still used because we want to clear the data in two tables after every test case. This is because different test cases write different data in to the test tables, and they need to be cleared before the next test case begins.